### PR TITLE
Add guardrails preventing complex params in LBFGS & SparseAdam

### PR DESCRIPTION
**Pull request by janeyx99**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118161
* #118160

### CODE REVIEW

File: torch/nn/parameter.py [hunk 1]
```diff
@@ -100,6 +100,7 @@ class UninitializedTensorMixin:
         torch.Tensor.to,
         torch.Tensor.get_device,
         torch._has_compatible_shallow_copy_type,
+        torch.is_complex,
```
Comment (mikaylagawarecki) [inline torch/nn/parameter.py:103]: nit: not a big deal but `torch.Tensor.is_complex` since the list is called `_allowed_methods`? 😛

**Review by mikaylagawarecki — APPROVED** [review]

File: torch/nn/parameter.py (hunk 1 above)
Comment (janeyx99) [reply torch/nn/parameter.py:103]: so is_complex is only a torch function...I had tried torch.Tensor.is_complex before and it didn't work :/

Comment (janeyx99) [reply torch/nn/parameter.py:103]: nah i am a troll, i was just using the torch equiv instead

**Pull request closed without merging**